### PR TITLE
WEB-2740: Adding support for "All" locales

### PR DESCRIPTION
**Pull request by ghost**
@bigcommerce/atx-web 

https://jira.bigcommerce.com/browse/WEB-2740

This improves upon the "Locales" functionality by adding in an "All" option for a redirect.

"All" works for all locales, except where there is a redirect with the same source, but for a specific locale. 

example:

`/old-busted/ -> /new-hotness/  - ALL` - will work for en_us, en_au
`/old-busted/ -> /new-hotness-UK/  - en_gb` - will only work for en_gb

Also, this only works for static redirects ATM. Entry ones will need their own brainstorming sessions to make sure all of the details are hammered out.

Testing:
This thing is a beast.  I think the easiest way is to checkout this branch, and then copy it in into Craft and run through the code that way

1. Go to https://bc.craft.local/hub/settings and clear the cache. **If you run into any trouble, try doing this before counting it as an error**
1. Make a new static redirect. Set the locale to whatever
1. Make sure this redirect works
1. Go back and try to make a duplicate redirect. It should stop you.
1. Go back and make a redirect with the same source and destination, but a different locale. It should allow this
1. Verify that redirect also works on the different locale
1. Make a redirect for "all". Verify that it works across all locales
1. Make a redirect with the same source as the one you just made, but a different destination. Set the locale to a specific instance
1. Verify that the redirect for "all" works on all of the other locales than the one you just set
1. Verify that the redirect for the specific locale works
1. Verify that any of the entry specific redirects are ignoring locales (they can be saving the locale in the DB, but we don't really care about the rest at the moment)

The migration style that has been used for this plugin is not as verbose as we usually do ours, but you can check the DB to see if it ran. Your indexes should look like

image

that `..._unq_idx` should have lines for both `redirectSrcUrlParsed` and `locale`

### CODE REVIEW

File: services/RetourService.php [hunk 1]
```diff
@@ -108,14 +113,56 @@ public function findRedirectMatch($url)
...
         if ($result) {
+            if ($result['locale'] == 'all' && !$this->checkCachedRedirectsForEachLocale($url)) {
+                //There's a chance that it found an "all" but
+                //there's also at least one specific match still lurking...
+                //cache those as well, and if it turns out it's actually the
+                //redirect we wanted, use it instead.
+                $query = craft()->db->createCommand()
+                    ->select('*')
+                    ->from('retour_static_redirects')
+                    ->where('redirectSrcUrl = "' . $url . '"')
```
Comment (tekstrand) [inline services/RetourService.php]: I have concerns about doing this here, I feel we're opening ourselves up to injection

**Review by tekstrand — CHANGES_REQUESTED** [review]

File: services/RetourService.php (hunk 1 above)
Comment (ghost) [reply services/RetourService.php]: Oh yeah great call!

I was so happy this _seemed_ to be working, I didn't think about that aspect.

I can sanitize that pretty easily (I think 🤔 ) and I'll re-up when I get it 🙇 

(It was also pretty late, so who know what I actually wrote 😛  )

Comment (ghost) [reply services/RetourService.php]: Easy! https://www.yiiframework.com/doc/guide/2.0/en/db-dao#binding-parameters

♻️

Comment (tekstrand) [reply services/RetourService.php]: 👏 thanks!

**Review by tekstrand — APPROVED** [review]
The functionality worked for me here. I still believe we could work to clean this code up but do _not_ believe it it worth our time at this point in the project. Would you mind creating a JIRA ticket for us to revisit both cleaning this up and supporting dynamic redirects?

**Pull request merged**